File: ProblemSolve/337.cpp
/********************
 * House Robber III *
 ********************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    //pair<현재 노드 선택, 현재 노드 선택 x> 값 리턴
    pair<int, int> calMax(TreeNode* root) {
        if(root == NULL){
            return {0, 0};
        }
        
        pair<int, int> left = calMax(root -> left);     //left subtree를 계산
        pair<int, int> right = calMax(root -> right);   //right subtee를 계산
        
        //현재 node를 선택하는 경우에는 child를 선택할 수 없다
        int choose = root -> val + left.second + right.second;
        
        //현재 node를 선택하지 않는 경우에는 child도 선택하지 않을 수 있으므로 child를 선택하는 경우, 선택하지 않는 경우 중 큰 값을 택한다
        int noChoose = max(left.first , left.second) + max(right.first, right.second);
        
        return {choose, noChoose};
    }
    
    int rob(TreeNode* root) {
        return max(calMax(root).first, calMax(root).second);
    }
};

File: ProblemSolve/337.cs
/********************
 * House Robber III *
 ********************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int Max(int a, int b){
        return a >= b ? a : b;
    }
    //DFS 방식으로 트리 순회
    public int[] CalMaxMoney(TreeNode root){
        if(root == null){
            return new int[2];
        }
        
        int[] left = CalMaxMoney(root.left);
        int[] right = CalMaxMoney(root.right);
        
        int choose = root.val + left[1] + right[1]; //현재 노드를 훔친 경우. 다음 노드는 훔치지 못함.
        int noChoose = Max(left[0], left[1]) + Max(right[0], right[1]); //현대 노드를 훔치지 않는 경우. 
                                                                        //다음 노드를 훔쳤을 때와 훔치지 않았을 때 중 더 큰 값 선택        
        return new int[2]{choose, noChoose};
    }
    
    public int Rob(TreeNode root) {
        int[] ansArr = CalMaxMoney(root);
        
        return Max(ansArr[0], ansArr[1]);
    }
}